### ISSUE TITLE
[SPEC] Internal Primary Key

### ISSUE THREAD

**Issue by KEINOS** [KEINOS/KVSQLite3#2]
### TL; DR

By default, this Class will convert the requested `key` string into a hashed 8 Byte (64 bit) integer to use it as a `PRIMARY KEY` internally.

- Current Algorithm: `sha512` and `md5`
- Basic Formula:

    ```php
    <?PHP

    $hash1 = hash('sha512', $key_request);
    $hash2 = hash('md5', $key_request);
    
    $len   = 8; // Character length of 4 Byte HEX string
    $hash1 = substr($hash1, 0, $len);
    $hash2 = substr($hash2, 0, $len);
    
    // 8 + 8 = 16 char = 8 Byte Hex String -> dec -> int
    $key_to_use = intval(hexdec($hash1 . $hash2));
    ```

### TS; DR

The concept of this formula is based on the following article:

- REFS:
  - [長さを知ったところで何？](https://qiita.com/KEINOS/items/c92268386d265042ea16#tsdr) | hashアルゴリズムとハッシュ値の長さ一覧 @ Qiita

In SQLite3, searching for a record with a specific `rowid` is FAST.

> Searching for a record with a specific `rowid`, or for all records with `rowid`s within a specified range is around **twice as fast** as a similar search made by specifying any other `PRIMARY KEY` or indexed value.
>
> <sup>(From: [ROWIDs and the INTEGER PRIMARY KEY](https://www.sqlite.org/lang_createtable.html#rowid) | SQL As Understood By SQLite」@ SQLite.org)</sup>

`rowid` is a 64-bit integer, which is 8 Byte, 16 characters in `HEX` string.

> Except for `WITHOUT ROWID` tables, all rows within **SQLite tables have a 64-bit signed integer key** that uniquely identifies the row within its table. This integer is usually called the "`rowid`". 
>
> <sup>(From: [ROWIDs and the INTEGER PRIMARY KEY](https://www.sqlite.org/lang_createtable.html#rowid) | SQL As Understood By SQLite」@ SQLite.org)</sup>

And if the primary key is defined as `INTEGER PRIMARY KEY` then the primary key becomes an alias of the `rowid`.

> If a table has the primary key that consists of one column, and that column defined as `INTEGER`, exactly INTEGER in any cases, such as, INTEGER, integer, etc., then **this primary key column becomes an *alias* for the `rowid` column**.
> 
> <sup>(From: [SQLite Primary Key](http://www.sqlitetutorial.net/sqlite-primary-key/) @ SQLiteTutorial.net)</sup>

Therefore, if we define the primary key as `INTEGER` then **we can search as fast as searching with the `rowid`**.

So hashing the key string into an integer, it will speed up searching the data.

```php
<?PHP
# Sample key requested
$key_original = 'my_key_to_store';

# Hash the key with 64 bit length "fnv1" hash algorythm
$hash_string = hash('fnv164', $key_original);

# Convert HEX string into DEC
$hash_dec = hexdec($hash_string);

# Make it sure to be an Integer
$hash_int = intvar($hash_dec);
...
```

Even though [`fnv1` is fast enough](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function), to avoid a collision little more, we'll combine two algorithms. Since there is much less probability that two different algorithms collide at the same time.

```php
<?php

functionget16ByteHash(string $value): string
{
    $len = 8; // Length of 4 Byte HEX string

    $hash1 = substr(hash('sha512', $value), 0, $len);
    $hash2 = substr(hash('md5', $value), 0, $len);

    // 4 + 4 Byte = 8 Byte = 64 bit = 16 chars in HEX String
    return hexdec($hash1 . $hash2);
}
```
